nemo/collections/asr/modules/multi_head_attention.py

```diff
@@ -0,0 +1,292 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Part of this code is adopted from https://github.com/espnet/espnet
+"""
+
+import math
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+__all__ = [
+    'RelPositionMultiHeadAttention',
+    'RelPositionalEncoding',
+    'PositionalEncoding',
+]
+
+
+class MultiHeadAttention(nn.Module):
+    """Multi-Head Attention layer.
+    Args:
+        n_head (int): number of heads
+        n_feat (int): size of the features
+        dropout_rate (float): dropout rate
+    """
+
+    def __init__(self, n_head, n_feat, dropout_rate):
+        """Construct an MultiHeadedAttention object."""
+        super(MultiHeadAttention, self).__init__()
+        assert n_feat % n_head == 0
+        # We assume d_v always equals d_k
+        self.d_k = n_feat // n_head
+        self.h = n_head
+        self.linear_q = nn.Linear(n_feat, n_feat)
+        self.linear_k = nn.Linear(n_feat, n_feat)
+        self.linear_v = nn.Linear(n_feat, n_feat)
+        self.linear_out = nn.Linear(n_feat, n_feat)
+        self.attn = None
+        self.dropout = nn.Dropout(p=dropout_rate)
+
+    def forward_qkv(self, query, key, value):
+        """Transforms query, key and value.
+        Args:
+            query (torch.Tensor): (batch, time1, size)
+            key (torch.Tensor): (batch, time2, size)
+            value (torch.Tensor): (batch, time2, size)
+        returns:
+            q (torch.Tensor): (batch, head, time1, size)
+            k (torch.Tensor): (batch, head, time2, size)
+            v (torch.Tensor): (batch, head, time2, size)
+        """
+        n_batch = query.size(0)
+        q = self.linear_q(query).view(n_batch, -1, self.h, self.d_k)
+        k = self.linear_k(key).view(n_batch, -1, self.h, self.d_k)
+        v = self.linear_v(value).view(n_batch, -1, self.h, self.d_k)
+        q = q.transpose(1, 2)
+        k = k.transpose(1, 2)
+        v = v.transpose(1, 2)
+
+        return q, k, v
+
+    def forward_attention(self, value, scores, mask):
+        """Compute attention context vector.
+        Args:
+            value (torch.Tensor): (batch, time2, size)
+            scores(torch.Tensor): (batch, time1, time2)
+            mask(torch.Tensor): (batch, time1, time2)
+        returns:
+            value (torch.Tensor): transformed `value` (batch, time2, d_model) weighted by the attention scores
+        """
+        n_batch = value.size(0)
+        if mask is not None:
+            mask = mask.unsqueeze(1).eq(0)  # (batch, 1, time1, time2)
+            min_value = float(np.finfo(torch.tensor(0, dtype=scores.dtype).numpy().dtype).min)
```

Comment:
Creating a torch tensor only to analyse the min possible float value? Can this be cached to preserve computation and allocations?

Reply:
It is just a single value tensor and needs to be as the same type of scores which are available in forward.

Reply:
Right I know its just one alloc, but this is a submodule in a larger network - so for 1 batch, it gets called num_layers number of times, and we have num_layers * dataset_len * num epochs / num_batches number of calls (in the naive case). So that adds up to millions of calls for longer runs. 

I think its fine, just the pytorch allocator is gonna have a lot of garbage cleaning to do maybe. But this logic - 
```python
torch.tensor(0, dtype=scores.dtype).numpy().dtype
```

is it not equivalent to
```python
if scores.dtype == torch.float16:
  dtype = np.float16
else:
  dtype = np.float32
np.finfo(dtype)
```

Reply:
done